### PR TITLE
Backport ModelRegistry Component

### PR DESCRIPTION
**Pull request by ykaliuta**
Jira: https://issues.redhat.com/browse/RHOAIENG-4272

### CODE REVIEW

**Comment by ykaliuta** [comment]
From the log https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/red-hat-data-services_rhods-operator/329/pull-ci-red-hat-data-services-rhods-operator-main-rhods-operator-e2e/1828679610703810560/artifacts/rhods-operator-e2e/gather-extra/artifacts/pods/redhat-ods-operator_redhat-ods-operator-controller-manager-d5f8fbcbd-dmfq4_manager.log

```{"level":"error","ts":"2024-08-28T07:13:25Z","logger":"opendatahub.controllers.DSCInitialization","msg":"error to set networkpolicy in operator namespace","path":"/opt/manifests/monitoring/networkpolicy","error":"networkpolicies.networking.k8s.io \"redhat-ods-operator\" is forbidden: cannot set blockOwnerDeletion in this case because cannot find RESTMapping for APIVersion dscinitialization.opendatahub.io/v1 Kind DSCInitialization: no matches for kind \"DSCInitialization\" in version \"dscinitialization.opendatahub.io/v1\"","stacktrace":"github.com/opendatahub-io/opendatahub-operator/v2/controllers/dscinitialization.(*DSCInitializationReconciler).reconcileDefaultNetworkPolicy\n\t/workspace/controllers/dscinitialization/utils.go:202\ngithub.com/opendatahub-io/opendatahub-operator/v2/controllers/dscinitialization.(*DSCInitializationReconciler).createOdhNamespace\n\t/workspace/controllers/dscinitialization/utils.go:121\ngithub.com/opendatahub-io/opendatahub-operator/v2/controllers/dscinitialization.(*DSCInitializationReconciler).Reconcile\n\t/workspace/controllers/dscinitialization/dscinitialization_controller.go:180\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235"}```
{"level":"error","ts":"2024-08-28T07:15:34Z","logger":"opendatahub.controllers.DSCInitialization","msg":"failed applying service mesh resources","error":"2 errors occurred:\n\t* failed applying FeatureHandler features. cause: 1 error occurred:\n\t* failed to get resource istio-system/data-science-smcp: no matches for kind \"ServiceMeshControlPlane\" in version \"maistra.io/v2\"\n\n\n\t* failed applying FeatureHandler features. cause: 1 error occurred:\n\t* 2 errors occurred:\n\t* failed to find Service Mesh Control Plane: no matches for kind \"ServiceMeshControlPlane\" in version \"maistra.io/v2\"\n\t* service mesh control plane is not ready\n\n\n\n\n\n","stacktrace":"github.com/opendatahub-io/opendatahub-operator/v2/controllers/dscinitialization.(*DSCInitializationReconciler).configureServiceMesh\n\t/workspace/controllers/dscinitialization/servicemesh_setup.go:44\ngithub.com/opendatahub-io/opendatahub-operator/v2/controllers/dscinitialization.(*DSCInitializationReconciler).Reconcile\n\t/workspace/controllers/dscinitialization/dscinitialization_controller.go:296\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/opt/app-root/src/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235"}```

Sounds weird
```

**Comment by ykaliuta** [comment]
Finally e2e passed :)

**Comment by VaishnaviHire** [comment]
@dhirajsb Should we go ahead with adding Model registry as component for 2.14?

**Comment by zdtsw** [comment]
> @dhirajsb Should we go ahead with adding Model registry as component for 2.14?

should we wait for a weeek before merge this PR?
we can have downstream nightly on pure webhook first and give odh nightly one week to test modelreg.

**Comment by dhirajsb** [comment]
I'll leave it up to platform team. We (MR and Dashboard teams) can test on nightly builds and 2.18 release this week. 
@tonyxrmdavidson cc

**Comment by ykaliuta** [comment]
I see the operator pod in the artefacts https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/red-hat-data-services_rhods-operator/329/pull-ci-red-hat-data-services-rhods-operator-main-rhods-operator-e2e/1836434864258158592/artifacts/rhods-operator-e2e/gather-extra/artifacts/pods/redhat-ods-operator_redhat-ods-operator-controller-manager-6d8cff779-8zx5p_manager.log . Wondering what the test cannot detect

File: README.md
Comment (zdtsw) [inline README.md:330]:
```suggestion
      registriesNamespace: "rhoai-model-registries"
```

File: README.md
Comment (zdtsw) [inline README.md:338]:
```suggestion
      managementState: Removed
```

File: bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (zdtsw) [inline bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:431]:
```suggestion
                        default: rhoai-model-registries
```

File: bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (zdtsw) [inline bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:434]:
```suggestion
                          to "rhoai-model-registries"
```

File: bundle/manifests/rhods-operator.clusterserviceversion.yaml
Comment (zdtsw) [inline bundle/manifests/rhods-operator.clusterserviceversion.yaml:51]:
```suggestion
                "registriesNamespace": "rhoai-model-registries"
```

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [inline components/modelregistry/modelregistry.go:33]:
```suggestion
	DefaultModelRegistriesNamespace = "rhoai-model-registries"
```

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [inline components/modelregistry/modelregistry.go:55]:
```suggestion
	// +kubebuilder:default="rhoai-model-registries"
```

File: config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (zdtsw) [inline config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml:447]:
```suggestion
                        default: rhoai-model-registries
```

File: config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (zdtsw) [inline config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml:450]:
```suggestion
                          to "rhoai-model-registries"
```

File: config/samples/datasciencecluster_v1_datasciencecluster.yaml
Comment (zdtsw) [inline config/samples/datasciencecluster_v1_datasciencecluster.yaml:45]:
```suggestion
      managementState: "Removed"
      registriesNamespace: "rhoai-model-registries"
```

File: docs/DESIGN.md
Comment (zdtsw) [inline docs/DESIGN.md:67]:
```suggestion
            registriesNamespace: "rhoai-model-registries"
```

File: docs/api-overview.md
Comment (zdtsw) [inline docs/api-overview.md:254]:
```suggestion
| `registriesNamespace` _string_ | Namespace for model registries to be installed, configurable only once when model registry is enabled, defaults to "rhoai-model-registries" | rhoai-model-registries | MaxLength: 63 <br />Pattern: `^([a-z0-9]([-a-z0-9]*[a-z0-9])?)?$` <br /> |
```

File: docs/upgrade-testing.md
Comment (zdtsw) [inline docs/upgrade-testing.md:141]:
```suggestion
      managementState: Removed
      registriesNamespace: "rhoai-model-registries"
```

File: get_all_manifests.sh
Comment (zdtsw) [inline get_all_manifests.sh:24]:
```suggestion
    ["modelregistry"]="red-hat-data-services:model-registry-operator:rhoai-2.14:config:model-registry-operator"
```

File: tests/e2e/helper_test.go
Comment (zdtsw) [inline tests/e2e/helper_test.go:173]:
```suggestion
						ManagementState: operatorv1.Removed,
```

File: README.md
Comment (ykaliuta) [reply README.md:338]: this hunk probably came with some automatic merge, 2 modelregistry entries

File: docs/upgrade-testing.md
Comment (ykaliuta) [reply docs/upgrade-testing.md:141]: this Managed -> Removed doc changes probably not necessary?

File: tests/e2e/helper_test.go
Comment (ykaliuta) [reply tests/e2e/helper_test.go:173]: So, just disable it in the e2e test, right?

**Comment by ykaliuta** [comment]
I'll do default namespace change by replacing in the diff directly. So on top changing branch for manifests and disabling in the e2e test. Right?

File: tests/e2e/helper_test.go
Comment (zdtsw) [reply tests/e2e/helper_test.go:173]: yes, dont think we need to set it as default in e2e, since it is not GA yet

File: docs/upgrade-testing.md
Comment (ykaliuta) [reply docs/upgrade-testing.md]: ok. Fine for me to change, but I wanted to keep difference with ODH as less as possible.

File: bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (zdtsw) [inline bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:431]: the biggest part i do not like is this one:  ( i know i am the one suggested for this value i this PR, however i am debating with myself).

@lburgazzoli  @VaishnaviHire .
if we use odh-model-registries as default value here,  we have the same API for odh and rhoai, which should be the case.
But we MIGHT have a problem: user gets a wrong idea of what is the value will be used out of the box or by reading CRD spec.
we can set the rhoai-model-regiestries in "sample" and CSV, still if anyone wants to create DSC CR via cli, it can be "odh-model-registeries" pass down to dashboard.

on the other hand, we support config different namespace. so if someone explicitly set "odh-model-registries" for a rhoai offering, it is acceptable too. 

:thinking: :thinking:  :thinking:

Comment (ykaliuta) [reply bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:431]: Can it use some platform independent name for both versions?

Comment (zdtsw) [reply bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:431]: i think it was a decision from modelreg team, not we want them to have different name in upstream and downstream.

File: config/samples/datasciencecluster_v1_datasciencecluster.yaml
Comment (zdtsw) [inline config/samples/datasciencecluster_v1_datasciencecluster.yaml:44]:
```suggestion
      managementState: "Removed"
```

File: bundle/manifests/rhods-operator.clusterserviceversion.yaml
Comment (zdtsw) [inline bundle/manifests/rhods-operator.clusterserviceversion.yaml:50]:
```suggestion
                "managementState": "Removed"
```

File: bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml
Comment (zdtsw) [reply bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:431]: sorry, i was stupid
this should not be a problem.
in API, the default can be the same in ODH and rhoai, "odh-model-regiestries" all we need to do is a different one in sample.

so here should be 
`default: odh-model-registries`

Comment (ykaliuta) [reply bundle/manifests/datasciencecluster.opendatahub.io_datascienceclusters.yaml:431]: No problem. All the entries?

**Review by zdtsw — APPROVED** [review]
once the default is revertd in DSC CR, ready to merge

File: tests/e2e/helper_test.go
Comment (dhirajsb) [reply tests/e2e/helper_test.go]: No, we need it enabled in the e2e test. Otherwise how will we know that the component reconcile works in rhoai 2.14?

**Comment by dhirajsb** [comment]
@zdtsw the default namespace in the CRD and manifests has to be `rhoai-model-registries`.

File: components/modelregistry/modelregistry.go
Comment (dhirajsb) [inline components/modelregistry/modelregistry.go:54]:
```suggestion
	// Namespace for model registries to be installed, configurable only once when model registry is enabled, defaults to "rhoai-model-registries"
```

File: config/samples/datasciencecluster_v1_datasciencecluster.yaml
Comment (dhirajsb) [inline config/samples/datasciencecluster_v1_datasciencecluster.yaml:45]:
```suggestion
      registriesNamespace: "rhoai-model-registries"
```

File: tests/e2e/helper_test.go
Comment (dhirajsb) [inline tests/e2e/helper_test.go:173]:
```suggestion
						ManagementState: operatorv1.Managed,
```

Comment (dhirajsb) [reply tests/e2e/helper_test.go:173]: Enable testing model registry reconcile in e2e tests.

**Review by dhirajsb — CHANGES_REQUESTED** [review]
Added more changes to use rhoai MR namespace `rhoai-model-registries` and to enable testing model registry reconcile.

File: components/modelregistry/modelregistry.go
Comment (zdtsw) [reply components/modelregistry/modelregistry.go:54]: i think the reason we keep it as odh-* is to ensure the source code is the same for upstream and downstream. to only config it with different values from csv etc

Comment (ykaliuta) [reply components/modelregistry/modelregistry.go:54]: that could be just revert to the previous version instead of all that suggestions.

**Comment by VaishnaviHire** [comment]
@dhirajsb @ykaliuta We probably need to update the api-docs here -

```
make generate manifests api-docs
```

**Comment by ykaliuta** [comment]
> @dhirajsb @ykaliuta We probably need to update the api-docs here -
> 
> ```
> make generate manifests api-docs
> ```

So, it now
```
 % git grep odh-model-registries
components/modelregistry/modelregistry.go:      // Namespace for model registries to be installed, configurable only once when model registry is enabled, defaults to "odh-model-registries"
docs/upgrade-testing.md:      registriesNamespace: "odh-model-registries"

```

After api update it is 
```
% git grep odh-model-registries   
components/modelregistry/modelregistry.go:      // Namespace for model registries to be installed, configurable only once when model registry is enabled, defaults to "odh-model-registries"
config/crd/bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml:                          to "odh-model-registries"
docs/api-overview.md:| `registriesNamespace` _string_ | Namespace for model registries to be installed, configurable only once when model registry is enabled, defaults to "odh-model-registries" | rhoai-model-registries | MaxLength: 63 <br />Pattern: `^([a-z0-9]([-a-z0-9]*[a-z0-9])?)?$` <br /> |
docs/upgrade-testing.md:      registriesNamespace: "odh-model-registries"
```

I already lost what modelregistry want and were, so @dhirajsb , please update it up to your requirements.

**Comment by ykaliuta** [comment]
> @dhirajsb @ykaliuta We probably need to update the api-docs here -
> 
> ```
> make generate manifests api-docs
> ```

forget my last comment. I updated the 2 entries I had if it's what is required  (will revert, if not). But about the generated files, the tree is clean after the command.

File: config/monitoring/prometheus/apps/prometheus-configs.yaml
Comment (dhirajsb) [inline config/monitoring/prometheus/apps/prometheus-configs.yaml:1658]:
```suggestion
            triage: "https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/blob/main/RHODS/Model-Serving/rhoai-model-registry-operator-probe-success-burn-rate.md"
```

File: config/monitoring/prometheus/apps/prometheus-configs.yaml
Comment (dhirajsb) [inline config/monitoring/prometheus/apps/prometheus-configs.yaml:1671]:
```suggestion
            triage: "https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/blob/main/RHODS/Model-Serving/rhoai-model-registry-operator-probe-success-burn-rate.md"
```

File: config/monitoring/prometheus/apps/prometheus-configs.yaml
Comment (dhirajsb) [inline config/monitoring/prometheus/apps/prometheus-configs.yaml:1684]:
```suggestion
            triage: "https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/blob/main/RHODS/Model-Serving/rhoai-model-registry-operator-probe-success-burn-rate.md"
```

**Review by dhirajsb — CHANGES_REQUESTED** [review]
Need to change the model registry SOP doc url.

**Review by zdtsw — APPROVED** [review]

**Pull request merged**
